src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Legacy/RoslynCSharpTokenizer.cs

```diff
@@ -386,404 +210,472 @@ private StateResult Data()
+        var curPosition = Source.Position;
+        var result = GetNextResult(NextResultType.Token);
+        var csharpToken = result.Token;
+        // Don't include trailing trivia; we handle that differently than Roslyn
+        var finalPosition = curPosition + csharpToken.Span.Length;
+
+        for (; curPosition < finalPosition; curPosition++)
         {
-            return HexLiteral();
+            TakeCurrent();
         }
-        else
+
+        Buffer.Clear();
+        _currentCSharpTokenTriviaEnumerator = (csharpToken.TrailingTrivia.GetEnumerator(), isLeading: false);
+        return Transition(RoslynCSharpTokenizerState.TriviaForCSharpToken, EndToken(csharpToken.Text, SyntaxKind.NumericLiteral));
+    }
+
+    private StateResult Identifier()
+    {
+        var curPosition = Source.Position;
+        var result = GetNextResult(NextResultType.Token);
+        var csharpToken = result.Token;
+        // Don't include trailing trivia; we handle that differently than Roslyn
+        var finalPosition = curPosition + csharpToken.Span.Length;
+
+        for (; curPosition < finalPosition; curPosition++)
         {
-            return DecimalLiteral();
+            TakeCurrent();
         }
+
+        var type = SyntaxKind.Identifier;
+        if (!csharpToken.IsKind(CSharpSyntaxKind.IdentifierToken) || result.ContextualKind is not (CSharpSyntaxKind.None or CSharpSyntaxKind.IdentifierToken))
+        {
+            type = SyntaxKind.Keyword;
+        }
+
+        var token = EndToken(csharpToken.Text, type);
+
+        Buffer.Clear();
+        _currentCSharpTokenTriviaEnumerator = (csharpToken.TrailingTrivia.GetEnumerator(), isLeading: false);
+        return Transition(RoslynCSharpTokenizerState.TriviaForCSharpToken, token);
     }
 
-    private StateResult HexLiteral()
+    private StateResult Transition(RoslynCSharpTokenizerState state, SyntaxToken? result)
     {
-        TakeUntil(c => !IsHexDigit(c));
-        TakeIntegerSuffix();
-        return Stay(EndToken(SyntaxKind.IntegerLiteral));
+        return Transition((int)state, result);
     }
 
-    private StateResult DecimalLiteral()
+    internal static CSharpSyntaxKind GetTokenKeywordKind(SyntaxToken token)
     {
-        TakeUntil(c => !Char.IsDigit(c));
-        if (CurrentCharacter == '.' && Char.IsDigit(Peek()))
-        {
-            return RealLiteral();
-        }
-        else if (IsRealLiteralSuffix(CurrentCharacter) ||
-                 CurrentCharacter == 'E' || CurrentCharacter == 'e')
+        if (token is null)
         {
-            return RealLiteralExponentPart();
+            return CSharpSyntaxKind.None;
         }
-        else
+
+        var content = token.Content;
+        return SyntaxFacts.GetKeywordKind(content) is var kind and not CSharpSyntaxKind.None
+            ? kind
+            : SyntaxFacts.GetContextualKeywordKind(content);
+    }
+
+    internal override CSharpSyntaxKind? GetTokenKeyword(SyntaxToken token)
```

Comment:
Should this simply call the static method above?

Reply:
The static method actually isn't used, so I'll just remove it.